KEP-4944: Kustomize Wasm Plugin (#4946)
kubernetes/enhancements: keps/sig-cli/4944-kustomize-wasm-plugin/README.md

The current Kustomize plugin ecosystem faces several operational challenges that
limit its effectiveness in various environments. Container plugins, while
providing good isolation, require a Docker runtime, which is often unavailable
or restricted in CI/CD environments due to security policies or resource
constraints. Exec plugins are tightly coupled to their underlying operating
system and execution environment, requiring platform-specific compilation and
distribution. This makes maintaining consistency across diverse environments
challenging. Additionally, container plugins introduce architecture-specific
dependencies, requiring builds for different CPU architectures (e.g., amd64,
arm64), which complicates deployment across heterogeneous infrastructure. These
limitations create scenarios where neither exec nor container plugins can be
effectively utilized, particularly in organizations with diverse infrastructure
requirements or strict security policies.

### Goals

1. Implement Wasm plugin support in the Kustomize KRM framework
1. Provide a secure sandboxed environment for executing custom transformations
1. Improve portability of custom transformers across different platforms

> 1. Enhance performance of resource transformations through WebAssembly

Reviewer: Nowhere in this document there's a clear information how this new mechanism will affect the kustomize version that is available in kubectl? IIRC the KRM functions are NOT available in kubectl, so I believe this one is also not going to be available, is that right assumption?

Author: Honesty, Exec KRM functions is disabled by kubectl, but the Container KRM functions is available for kustomize in kubectl now.
(We can use with `kubectl kustomize --enable-alpha-plugins ...`)
https://github.com/kubernetes-sigs/kustomize/issues/4556#issuecomment-1092095023

According to my understanding, the Exec KRM functions mechanism, which directly executes binaries, raises security concerns and was disabled in kubectl. With the Wasm Plugin, since the execution environment is sandboxed, I thought it seemed reasonable to make it available in the same way as current Container functions.

#### Story 2

A platform team wants to share team-specific resource transformation logic as
Kustomize plugins. By distributing Wasm plugins through OCI registries, team
members can easily obtain plugins and start using them without worrying about
execution environment setup or binary compatibility.

#### Story 3

An application developer wants to use a popular open-source Kustomize plugin
that implements advanced JSON patching, which is distributed through an OCI
registry. Since the plugin is implemented in WebAssembly and runs in a sandboxed
environment, they can confidently use the third-party plugin without worrying
about security vulnerabilities or system access. The developer can easily fetch
the latest version of the plugin from the OCI registry, verify its integrity,
and run it safely within their build pipeline, knowing that the plugin's
capabilities are limited to the intended resource transformations.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

> - Current kustomize plugin systen that called `Exec/Container KRM functions` is
>   using stdin/stdout to communicate with kustomize to plugins.

Reviewer: ```suggestion
- Current kustomize plugin system called `Exec/Container KRM functions` is
  using stdin/stdout to communicate between kustomize and the plugin.
```

- Now,
  [Wasm binary can be distributed with OCI registry](https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/).

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

1. WASI Specific is not yet GA. Any disruptive changes in the future will incur
   user transition costs.
1. The Wasm OCI Artifact specification is a relatively new standard in the OCI
   ecosystem. Maybe not mature enough.
1. Containaized KRM function is alpha feature but that used some user.
   - Provide container to exec conversion tools.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> Implement with [wazero](https://github.com/tetratelabs/wazero) to WASM runtime
> like the
> [kube-scheduler-wasm-extension](https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension)
> project.

Reviewer: ```suggestion
The implementation will use [wazero](https://github.com/tetratelabs/wazero) library for WASM runtime support. This is the same library used by [kube-scheduler-wasm-extension](https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension) project.
```

Reviewer: Additional question, there are several other libraries providing wasm support. Wasmer, Wasmtime and there's also native Go support. Can you describe in a few sentences why this, and not the other ones?

### Interface

Using the WASI to plugin interface.

##### with Wasm binary file

```yaml
apiVersion: transformers.example.co/v1
kind: App
metadata:
  name: example-app
  annotations:
    config.kubernetes.io/function: |
      wasip1:
        file: ../bin/fnapp.wasm
spec:
  port: 8080
```

##### with OCI registry

```yaml
apiVersion: transformers.example.co/v1
kind: App
metadata:
  name: example-app
  annotations:
    config.kubernetes.io/function: |
      wasip1:
        image: ghcr.io/koba1t/krm-fn-app:v0.0.3-wasm
spec:
  port: 8080

> ```

Reviewer: I'm not sure I follow, what the above files are meant for.

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[ ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes
necessary to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

> ##### Unit tests

Reviewer: For kustomize, I'm certain we can just skip interrogation and e2e tests, but listing units would be reasonable. 
Something like:
```
- sigs.k8s.io/kustomize/kustomize/v5/commands/build	- 2025-06-01 - 79.4
- sigs.k8s.io/kustomize/kustomize/v5/commands/create - 2025-06-26 - 84.9
etc...
```
Just make sure to list only the directories you'll be modifying as part of this functionality. There's no need to list all of them :wink:

These may be defined in terms of API maturity, [feature gate] graduations, or as
something else. The KEP should keep this high-level with a focus on what
signals will be looked at to determine graduation.

Consider the following in developing the graduation criteria for this enhancement:
- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
- [Feature gate][feature gate] lifecycle
- [Deprecation policy][deprecation-policy]

Clearly define what graduation means by either linking to the [API doc
definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
or by redefining what graduation means.

In general we try to use the same stages (alpha, beta, GA), regardless of how the
functionality is accessed.

[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

> #### Alpha

Reviewer: Clear graduation criteria will be needed. In the risks section you've mentioned that WASI is still not fully stable, is there a plan when it will be? Should we block GA promotion of this functionality until it does?
